#### **Step 1: Load the Dataset**

In [1]:
#read csv file from drive.
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [2]:
file_path = '/content/drive/MyDrive/Datasets/collection_with_abstracts.csv'
df = pd.read_csv(file_path)
df


PMID                                              Title  \
0      39435445  Editorial: The operationalization of cognitive...   
1      39398866  Characterization of arteriosclerosis based on ...   
2      39390053  Multi-scale input layers and dense decoder agg...   
3      39367648  An initial game-theoretic assessment of enhanc...   
4      39363262  Truncated M13 phage for smart detection of E. ...   
...         ...                                                ...   
11445  10607521  The characteristics of epidemics and invasions...   
11446  10072741  Effects of sales promotion on smoking among U....   
11447   9200018  Hypertension in an inner-city minority population   
11448   8039948  Aerosol transmission of a viable virus affecti...   
11449   1588926  Role of desolvation energy in the nonfacilitat...   

                                                 Authors  \
0      Winter M, Probst T, Tallon M, Schobel J, Pryss R.   
1      Zhou J, Li X, Demeke D, Dinh TA, Yang Y, Janow...   
2                                          Lan X, Jin W.   
3      Fatemi MY, Lu Y, Diallo AB, Srinivasan G, Azhe...   
4      Yuan J, Zhu H, Li S, Thierry B, Yang CT, Zhang...   
...                                                  ...   
11445               Cruickshank I, Gurney WS, Veitch AR.   
11446                                        Redmond WH.   
11447                                          Wieck KL.   
11448                    Grant RH, Scheidt AB, Rueff LR.   
11449                             Wright LL, Painter GR.   

                                                Citation   First Author  \
0      Front Neurosci. 2024 Oct 7;18:1501636. doi: 10...       Winter M   
1      J Med Imaging (Bellingham). 2024 Sep;11(5):057...         Zhou J   
2      Sci Rep. 2024 Oct 10;14(1):23729. doi: 10.1038...          Lan X   
3      Brief Bioinform. 2024 Sep 23;25(6):bbae476. do...      Fatemi MY   
4      J Nanobiotechnology. 2024 Oct 3;22(1):599. doi...         Yuan J   
...                                                  ...            ...   
11445  Theor Popul Biol. 1999 Dec;56(3):279-92. doi: ...  Cruickshank I   
11446  Prev Med. 1999 Mar;28(3):243-50. doi: 10.1006/...     Redmond WH   
11447  J Cardiovasc Nurs. 1997 Jul;11(4):41-9. doi: 1...       Wieck KL   
11448  Int J Biometeorol. 1994 May;38(1):33-9. doi: 1...       Grant RH   
11449              Mol Pharmacol. 1992 May;41(5):957-62.      Wright LL   

                     Journal/Book  Publication Year Create Date        PMCID  \
0                  Front Neurosci              2024  2024/10/22  PMC11491427   
1      J Med Imaging (Bellingham)              2024  2024/10/14  PMC11466048   
2                         Sci Rep              2024  2024/10/10  PMC11467340   
3                 Brief Bioinform              2024  2024/10/05  PMC11452536   
4             J Nanobiotechnology              2024  2024/10/04  PMC11451008   
...                           ...               ...         ...          ...   
11445            Theor Popul Biol              1999  1999/12/23          NaN   
11446                    Prev Med              1999  1999/03/12          NaN   
11447           J Cardiovasc Nurs              1997  1997/07/01          NaN   
11448           Int J Biometeorol              1994  1994/05/01          NaN   
11449               Mol Pharmacol              1992  1992/05/01          NaN   

      NIHMS ID                               DOI  \
0          NaN        10.3389/fnins.2024.1501636   
1          NaN         10.1117/1.JMI.11.5.057501   
2          NaN        10.1038/s41598-024-74701-0   
3          NaN               10.1093/bib/bbae476   
4          NaN        10.1186/s12951-024-02881-y   
...        ...                               ...   
11445      NaN            10.1006/tpbi.1999.1432   
11446      NaN            10.1006/pmed.1998.0410   
11447      NaN  10.1097/00005082-199707000-00005   
11448      NaN                10.1007/BF01241802   
11449      NaN 

In [3]:
print(df['Abstract'][12])
# df['Abstract'][1]

Computationally expensive data processing in neuroimaging research places demands on energy consumption-and the resulting carbon emissions contribute to the climate crisis. We measured the carbon footprint of the functional magnetic resonance imaging (fMRI) preprocessing tool fMRIPrep, testing the effect of varying parameters on estimated carbon emissions and preprocessing performance. Performance was quantified using (a) statistical individual-level task activation in regions of interest and (b) mean smoothness of preprocessed data. Eight variants of fMRIPrep were run with 257 participants who had completed an fMRI stop signal task (the same data also used in the original validation of fMRIPrep). Some variants led to substantial reductions in carbon emissions without sacrificing data quality: for instance, disabling FreeSurfer surface reconstruction reduced carbon emissions by 48%. We provide six recommendations for minimising emissions without compromising performance. By varying par

####**Step 2: Text preprocessing.**

select the necessary coloumns for further processing.


In [4]:
#selected the coloumns PMID, Title, Journal/Book, Publication Year, Abstract

selected_columns = df[['PMID', 'Title', 'Journal/Book', 'Abstract']]
selected_columns


PMID                                              Title  \
0      39435445  Editorial: The operationalization of cognitive...   
1      39398866  Characterization of arteriosclerosis based on ...   
2      39390053  Multi-scale input layers and dense decoder agg...   
3      39367648  An initial game-theoretic assessment of enhanc...   
4      39363262  Truncated M13 phage for smart detection of E. ...   
...         ...                                                ...   
11445  10607521  The characteristics of epidemics and invasions...   
11446  10072741  Effects of sales promotion on smoking among U....   
11447   9200018  Hypertension in an inner-city minority population   
11448   8039948  Aerosol transmission of a viable virus affecti...   
11449   1588926  Role of desolvation energy in the nonfacilitat...   

                     Journal/Book  \
0                  Front Neurosci   
1      J Med Imaging (Bellingham)   
2                         Sci Rep   
3                 Brief Bioinform   
4             J Nanobiotechnology   
...                           ...   
11445            Theor Popul Biol   
11446                    Prev Med   
11447           J Cardiovasc Nurs   
11448           Int J Biometeorol   
11449               Mol Pharmacol   

                                                Abstract  
0                                                    NaN  
1      PURPOSE: Our purpose is to develop a computer ...  
2      Accurate segmentation of COVID-19 lesions from...  
3      The application of deep learning to spatial tr...  
4      BACKGROUND: The urgent need for affordable and...  
...                                                  ...  
11445  In this paper we report the development of a h...  
11446  OBJECTIVE: The purpose of this study was to ex...  
11447  This study describes an inner-city elderly min...  
11448  A Gaussian diffusion model was applied to an e...  
11449  The ability of thymidine and four dideoxyribos...  

[11450 rows x 4 columns]

In [5]:
#checking null values.
selected_columns.isnull().sum()

PMID              0
Title             0
Journal/Book      0
Abstract        213
dtype: int64

To address the 213 missing values in the 'Abstract' column, I combined the 'Title' and 'Journal/Book' columns to fill these gaps.

This approach helps retain relevant information for each record, improving data consistency and completeness for subsequent tasks, such as classification and method extraction.

In [6]:
# Fill missing abstracts
selected_columns['Combined'] = selected_columns['Title'].fillna('') + ' ' + selected_columns['Journal/Book'].fillna('')
selected_columns['Abstract'] = selected_columns['Abstract'].fillna(selected_columns['Combined'])
selected_columns = selected_columns.drop('Combined', axis=1)
selected_columns.head()

<ipython-input-6-286c21ddfaae>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  selected_columns['Combined'] = selected_columns['Title'].fillna('') + ' ' + selected_columns['Journal/Book'].fillna('')
<ipython-input-6-286c21ddfaae>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  selected_columns['Abstract'] = selected_columns['Abstract'].fillna(selected_columns['Combined'])


PMID                                              Title  \
0  39435445  Editorial: The operationalization of cognitive...   
1  39398866  Characterization of arteriosclerosis based on ...   
2  39390053  Multi-scale input layers and dense decoder agg...   
3  39367648  An initial game-theoretic assessment of enhanc...   
4  39363262  Truncated M13 phage for smart detection of E. ...   

                 Journal/Book  \
0              Front Neurosci   
1  J Med Imaging (Bellingham)   
2                     Sci Rep   
3             Brief Bioinform   
4         J Nanobiotechnology   

                                            Abstract  
0  Editorial: The operationalization of cognitive...  
1  PURPOSE: Our purpose is to develop a computer ...  
2  Accurate segmentation of COVID-19 lesions from...  
3  The application of deep learning to spatial tr...  
4  BACKGROUND: The urgent need for affordable and...

In [7]:
#checking null values again.
selected_columns.isnull().sum()

PMID            0
Title           0
Journal/Book    0
Abstract        0
dtype: int64

####**Step 3: Data cleaning**
With all gaps filled, the next step is to clean the text fields. This involves removing stopwords, punctuation, and converting text to lowercase to standardize and streamline the data for further processing.


In [8]:
import nltk
nltk.download('stopwords')
nltk.download('punkt')
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
import string

stop_words = set(stopwords.words('english'))
punctuation = set(string.punctuation)

def clean_text(text):
  if isinstance(text, str):  # Check if the input is a string
    words = word_tokenize(text.lower())
    cleaned_words = [word for word in words if word not in stop_words and word not in punctuation]
    return " ".join(cleaned_words)
  else:
    return ""  # Or handle non-string values as needed

selected_columns['Title'] = selected_columns['Title'].apply(clean_text)
selected_columns['Journal/Book'] = selected_columns['Journal/Book'].apply(clean_text)
selected_columns['Abstract'] = selected_columns['Abstract'].apply(clean_text)

selected_columns.head()

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


PMID                                              Title  \
0  39435445  editorial operationalization cognitive systems...   
1  39398866  characterization arteriosclerosis based comput...   
2  39390053  multi-scale input layers dense decoder aggrega...   
3  39367648  initial game-theoretic assessment enhanced tis...   
4  39363262  truncated m13 phage smart detection e. coli da...   

               Journal/Book                                           Abstract  
0            front neurosci  editorial operationalization cognitive systems...  
1  j med imaging bellingham  purpose purpose develop computer vision approa...  
2                   sci rep  accurate segmentation covid-19 lesions medical...  
3           brief bioinform  application deep learning spatial transcriptom...  
4       j nanobiotechnology  background urgent need affordable rapid detect...

In [9]:
print(selected_columns['Abstract'][1])

purpose purpose develop computer vision approach quantify intra-arterial thickness digital pathology images kidney biopsies computational biomarker arteriosclerosis approach severity arteriosclerosis scored 0 3 753 arteries 33 trichrome-stained whole slide images wsis kidney biopsies outer contours media intima lumen manually delineated renal pathologist developed multi-class deep learning dl framework segmenting different intra-arterial compartments training dataset 648 arteries 24 wsis testing dataset 105 arteries 9 wsis subsequently employed radial sampling made measurements media intima thickness function spatially encoded polar coordinates throughout artery pathomic features extracted measurements collectively describe arterial wall characteristics technique first validated numerical analysis simulated arteries systematic deformations applied study effect arterial thickness measurements compared computationally derived measurements pathologists grading arteriosclerosis results num

 #### **Step 4: Semantic Filtering**
 Since the Abstract field provides the most detailed insight into each paper's content, it is used as the primary source for filtering.

 This method prioritizes contextual understanding to accurately capture papers that apply deep learning techniques to virology/epidemiology.



To achieve this, SBERT (Sentence-BERT) is used for semantic NLP filtering, enabling more accurate identification of relevant papers based on meaning rather than simple keyword matching.

In [32]:
# Install Necessary Libraries
!pip install -U sentence-transformers

#### **Load the SBERT Model**

Load a pretrained SBERT model from the sentence-transformers library.

For this semantic similarity task, all-MiniLM-L6-v2 is a good choice due to its efficiency and accurate embeddings, making it well-suited for filtering relevant papers.

In [10]:
from sentence_transformers import SentenceTransformer, util
# Load the SBERT model
#model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')
model = SentenceTransformer('all-MiniLM-L6-v2')
model


/usr/local/lib/python3.10/dist-packages/sentence_transformers/cross_encoder/CrossEncoder.py:13: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm, trange
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

SentenceTransformer(
  (0): Transformer({'max_seq_length': 256, 'do_lower_case': False}) with Transformer model: BertModel 
  (1): Pooling({'word_embedding_dimension': 384, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Normalize()
)

#### **Defining Target Terms for Semantic Filtering.**

These target terms provide a comprehensive set of phrases related to the application of deep learning in virology/epidemiology.
Using the SBERT model, reference embeddings are then generated for each target term. These embeddings serve as the basis for semantic similarity comparisons during filtering, helping to accurately identify relevant papers.

In [11]:
# Define main categories for combining terms dynamically
fields = ["virology", "epidemiology"]
techniques = [
    "deep learning", "neural network", "artificial neural network", "machine learning model","feedforward neural network","neural net algorithm",
    "convolutional neural network", "recurrent neural network", "LSTM", "multilayer perceptron",
    "CNN", "RNN", "GRNN", "computer vision", "vision models", "image processing","vision algorithms","computer graphics and vision",
    "object recognition", "scene understanding", "deep neural networks","computational semantics",
    "natural language processing", "text mining", "NLP", "computational linguistics","text analytics"
    "language modeling", "text analysis", "generative AI", "generative deep learning","text data analysis",
    "generative models", "generative artificial intelligence", "transformer models", "self-attention models",
    "transformer architecture", "attention-based neural networks", "transformer networks","speech and language technology",
    "sequence-to-sequence models", "large language model", "llm", "transformer-based model",
    "pretrained language model", "generative language model", "foundation model","language processing",
    "state-of-the-art language model", "multimodal model", "multimodal neural network",
    "vision transformer", "diffusion model", "generative diffusion model", "long short-term memory network",
    "diffusion-based generative model", "continuous diffusion model","textual data analysis"
]

# Combine terms
target_terms = [f"{tech} in {field}" for tech in techniques for field in fields]

# Generate reference embeddings
reference_embedding = model.encode(target_terms, convert_to_tensor=True)


#### **Generate Embeddings for Each Record's Abstracts**

To facilitate semantic filtering, embeddings are generated for each paper's 'Abstract' by applying the SBERT model.
These embeddings capture the semantic meaning of each 'Abstract', enabling accurate similarity comparisons with target terms.

In [12]:
# Generate embeddings for each paper's Abstract
selected_columns['abstract_embedding'] = selected_columns['Abstract'].apply(lambda x: model.encode(x, convert_to_tensor=True))
selected_columns.head()

PMID                                              Title  \
0  39435445  editorial operationalization cognitive systems...   
1  39398866  characterization arteriosclerosis based comput...   
2  39390053  multi-scale input layers dense decoder aggrega...   
3  39367648  initial game-theoretic assessment enhanced tis...   
4  39363262  truncated m13 phage smart detection e. coli da...   

               Journal/Book  \
0            front neurosci   
1  j med imaging bellingham   
2                   sci rep   
3           brief bioinform   
4       j nanobiotechnology   

                                            Abstract  \
0  editorial operationalization cognitive systems...   
1  purpose purpose develop computer vision approa...   
2  accurate segmentation covid-19 lesions medical...   
3  application deep learning spatial transcriptom...   
4  background urgent need affordable rapid detect...   

                                  abstract_embedding  
0  [tensor(0.0982, device='cuda:0'), tensor(-0.03...  
1  [tensor(-0.0169, device='cuda:0'), tensor(-0.0...  
2  [tensor(0.0098, device='cuda:0'), tensor(-0.06...  
3  [tensor(-0.0076, device='cuda:0'), tensor(-0.0...  
4  [tensor(-0.0285, device='cuda:0'), tensor(-0.0...

#### **Step 5. Filtering Papers Based on Semantic Similarity**
 Each paper’s Abstract embedding is compared to the reference embeddings generated from target terms.

 Using cosine similarity, a score is calculated for each paper, representing how closely it aligns with the target terms. Papers with a similarity score at or above the threshold are retained for further analysis.

In [13]:
# Set similarity threshold (adjust based on results)
similarity_threshold = 0.5

# Filter papers based on similarity score to reference embedding
selected_columns['similarity_score'] = selected_columns['abstract_embedding'].apply(
    lambda x: util.cos_sim(x, reference_embedding).max().item()
)
filtered_papers = selected_columns[selected_columns['similarity_score'] >= similarity_threshold]


####**Step 6. Export the Filtered Papers**
Once filtered, save the relevant papers to a new file for further processing or analysis.

In [14]:
filtered_papers.to_csv('filtered_deep_learning_virology_epidemiology.csv', index=False)


In [15]:
# Load the filtered deep learning papers in virology/epidemiology
df1 = pd.read_csv('filtered_deep_learning_virology_epidemiology.csv')
df1

PMID                                              Title  \
0    39013794  deep learning methods amplify epidemiological ...   
1    38473002  advancements glaucoma diagnosis role ai medica...   
2    38454859  scope artificial intelligence retinopathy prem...   
3    38424562  prevalence computer vision syndrome covid-19 p...   
4    37920033  unet segmentation network covid-19 ct images m...   
..        ...                                                ...   
725  16387332    diffusive si model allee effect application fiv   
726  16289268  dynamics dengue epidemics large-scale information   
727  11177527                          diffusion theory drug use   
728  10607521     characteristics epidemics invasions thresholds   
729   8039948  aerosol transmission viable virus affecting sw...   

            Journal/Book                                           Abstract  \
0            j epidemiol  deep learning subfield artificial intelligence...   
1      diagnostics basel  progress artificial intelligence algorithms di...   
2    indian j ophthalmol  artificial intelligence ai revolutionary techn...   
3      bmc public health  background computer vision syndrome become sig...   
4        math biosci eng  recent years global outbreak covid-19 posed ex...   
..                   ...                                                ...   
725          math biosci  minimal reaction-diffusion model spatiotempora...   
726     theor popul biol  model spatial temporal dynamics dengue fever p...   
727            addiction  paper examines applicability diffusion model d...   
728     theor popul biol  paper report development highly efficient nume...   
729    int j biometeorol  gaussian diffusion model applied epizootic pse...   

                                    abstract_embedding  similarity_score  
0    tensor([ 3.6120e-02, -2.2385e-02,  2.9091e-02,...          0.758788  
1    tensor([ 1.7732e-02, -3.1443e-02,  1.9534e-02,...          0.514860  
2    tensor([-6.3707e-02, -2.1677e-02, -2.9215e-02,...          0.535145  
3    tensor([ 6.4034e-02,  1.7856e-02, -1.1642e-02,...          0.521855  
4    tensor([ 5.6156e-02, -4.8756e-02,  2.3895e-02,...          0.532703  
..                                                 ...               ...  
725  tensor([ 4.0788e-02, -6.0069e-02,  1.2528e-02,...          0.501422  
726  tensor([ 8.1916e-02, -6.3010e-02,  3.9929e-02,...          0.511256  
727  tensor([ 1.2172e-01, -4.6752e-02, -3.3996e-02,...          0.588906  
728  tensor([ 1.6910e-02, -8.1633e-02,  2.1418e-02,...          0.586339  
729  tensor([ 7.4568e-02, -6.8140e-02,  1.5795e-02,...          0.541807  

[730 rows x 6 columns]

In [16]:
df1["Abstract"][0]

'deep learning subfield artificial intelligence machine learning based mostly neural networks often combined attention algorithms used detect identify objects text audio images video serghiou rough j epidemiol 0000 000 00 :0000-0000 present primer epidemiologists deep learning models models provide substantial opportunities epidemiologists expand amplify research data collection analyses increasing geographic reach studies including research subjects working large high dimensional data tools implementing deep learning methods quite yet straightforward ubiquitous epidemiologists traditional regression methods found standard statistical software exciting opportunities interdisciplinary collaboration deep learning experts epidemiologists statisticians healthcare providers urban planners professionals despite novelty methods epidemiological principles assessing bias study design interpretation others still apply implementing deep learning methods assessing findings studies used'